### ISSUE TITLE
Temporary Laser Mode Command

### ISSUE THREAD

**Issue by AustinSaintAubin** [gnea/grbl#406]
Would it be feasible to have a command to temporarily set [Laser Mode](https://github.com/gnea/grbl/wiki/Grbl-v1.1-Laser-Mode).
A command similar to Temporary Offset, or a specifying a special tool would be very useful.
```
; Temporary Offset
G92 X0 Y0 Z0  ; Set Temporary Offset
G92.1 X0 Y0 Z0  ; Unset Temporary Offset, reset of controller also clears the offset.

; Specify Tool
M6 T0  ; Tool #0
M6 T1  ; Tool #1
M6 T2  ; Tool #2
```

For machines like my'n that have a spindle and a laser, setting `$32=1` & `$32=0`, depending on the job, introduces a lot wear on the EEPROM.
With my machine I prefer to leave laser mode OFF / DISABLED, because having laser mode enabled by default can damage my machine if I where to forget to disable laser mode when carving.
Most of my laser scripts contain `$32=1` at the header, and `$32=0` at the footer.
This way when I carve with the spindle I am most often assured that GRBL is NOT in laser mode.

-----------

I recommend having I setting that will enable laser mode if a particular tool is specified.
```
$33=1 ; Laser Mode will enable if Tool #1 is selected (M6 T1), all other tools will have laser mode disabled by default. Other tool number could be specified to have laser mode enabled for.
```

Then laser mode could automatically enabled if tool #1 (or other tool # that is specified).
```
... other g-code
M6 T1 ; select laser tool.
M4 1; Laser Power
.... laser code.
... footer code.
end of job.
```

**Comment by AustinSaintAubin** [comment]
Alternatively, for `$32=1` could specify the tool laser mode is enabled on, instead of just enabling laser mode for 'all tools'.

**Comment by chamnit** [comment]
@AustinSaintAubin : I like this idea. However, there might be some special situations that may need to be accounted for when switching between the modes. But, this shouldn't be that difficult to work out.

In the next version, I think the laser enable setting can be used to disable laser mode completely and only when enabled, it can configured to denote all tools, a specific tool number, or a range of preset tools. I particularly like the latter because this would enable certain laser modes that could be used for multi materials or complex jobs that require different PWM frequencies or a tighter range of power levels. It's conducive to very fast switching of modes and CAM programs can easy generate and support this. I'll make sure to add this to my development notes for the next major release and add it there.

**Comment by AustinSaintAubin** [comment]
I like your latter idea as well.
As part of selecting the tool/laser mode, the RPM/PWM could be set.
This would be helpful when working with some software the expects to send a particular power range, like [LightBurn Software](https://lightburnsoftware.com/) ([0-1000](https://lightburn.fider.io/ideas/118/machine-laser-g-code-power-range))=[0-255]PWM, or a lot of other software (0-255)=[0-255]PWM.
Also of course the spindle with its expected RPM range (1200-2200)=[0-255] correlated to the PWM output, with laser mode disabled.

Having collectible modes / profiles would be awesome, but I leave all to your expertise. Thanks!

**Comment by bgort** [comment]
I ended up implementing 'temporary $32' mode myself for the same reason:  no need to wear the EEPROM when switching between laser and spindle in machines that have both.  My laser gcode starts/ends with $32=x, and for those with dual-purpose machines this seems reasonable.  For laser-dedicated machines, you'd probably just want to leave it.

FWIW, I like the idea using tools and configuring laser mode, PWM, etc. around them.  That would be much cleaner and seems like the better long term solution.  Perhaps this can be combined with the work I did (am doing) referenced here:  https://github.com/gnea/grbl-Mega/issues/69 ?

**Comment by quezadaminter** [comment]
@AustinSaintAubin , I am trying the same technique, I built a post-processor for Fusion 360 to deal with laser mode and working on having it auto-detect the laser mode toggle based on a tool number. But I am getting some inconsistent behavior from my Grbl controller. Here's a sample gcode output:

(1001)
G90 G94
G17
G21
;G43.1 Z0
(Enable the laser mode!)
(Using X offset: 0)
(Using Y offset: 0)
(Using Z offset: 0)
$32=1
G28 G91 Z0
G90

(2D Contour5)
S12000 M4
G54
**G0 X350.6 Y152.4**
Z15
Z0
G1 Z-0.2 F333.3
Y261.462 F5000
G3 X345.6 Y266.462 I-5
G1 X60.8
G3 X55.8 Y261.462 J-5
G1 Y43.338
G3 X60.8 Y38.338 I5
G1 X345.6
G3 X350.6 Y43.338 J5
G1 Y152.4
G0 Z0
G28 G91 Z0
G90
G28 G91 X0 Y0
G90
M5
;G49
(Disable the laser mode!)
$32=0

The odd behavior I am seeing is that the first time I send the code it works well, no errors reported. However, the next time I send it, Grbl seems to ignore the G0 instruction (in bold) above and brings the tool down to Z0 at the G28 XY location and attempts to start working from there. If I take out the $ commands for the laser then the Gcode works as expected.

What I am wondering is if you have seen this behavior before? What does your Gcode look like when you are inserting the $32 commands to manipulate the laser mode?

By the way, I built and installed grbl-sim to test with and I have no problems reported by it for the same file. Of course it doesn't currently give me position information so I can't tell if it is failing on the G0 as well in the sim.

I am using bCNC as a sender so I am not sure if the problem is with their parser. UGS refuses to send $ commands when Grbl is not in IDLE mode.

**Comment by AustinSaintAubin** [comment]
@quezadaminter, I can’t say for sure if I have experienced this exact issue. However, I have experienced inconsistencies when running some code the second time. I use CNCjs w/ GRBL. Have not run into this possible bug lately, so difficult to recall. I have usually issue a reset before running a job to clear out anything that I might have changed / tweaked during testing of a setup. 

Would you like me to test your code and see what the results are? Also, you might get a cheap  Arduino and install GRBL on it. I run commands on an Arduino Uno when I want to test dangerous or new things will zero risk. Also look at CNCjs.

**Comment by quezadaminter** [comment]
@AustinSaintAubin, yes, if you have time give the above code a try. It's a rectangle with rounded corners. If your controller is not set up to work safely with the G28 command then make sure you comment them out before trying to run it or you will crash the machine.

I am actually following your guide to install the RPi with CNCjs on my setup, just haven't gotten that far :). In the meantime I am writing this post processor to automate mixing a carving/milling operation with a laser operation from fusion. Eventually should be able to pick it up from a tool number like you suggested above. If we get a temporary parameter in Grbl it will be even more streamlined!

I do have a spare Arduino to test with, I'll try that, although the simulator works quite well and you can also test code on the machine with the $C flag. If you do try the code, let me know what you see happen, all I was doing is running the operation once, letting it finish and then hitting play again for a second run. In my case the second run would not throw an error, it would cut in the wrong place and I found like you said that a soft reset was the only way to get the controller back on track. I guess I could add a soft reset at the top of the gcode in the post processor, just seems dirty.

**Comment by AustinSaintAubin** [comment]
[GRBL Issue #406 for @quezadaminter ](https://youtu.be/IiofqGnkzNY)
[![IGRBL Issue #406 for @quezadaminter](https://img.youtube.com/vi/IiofqGnkzNY/0.jpg)](https://youtu.be/nQacH3seynM)

Tested your gcode a few times.
Issue is... consistently inconsistent, every other run does not finish.
However, fails at the same spot every other time.

```
G90 G94
G17
G21
;G43.1 Z0
;(Enable the laser mode!)
;(Using X offset: 0)
;(Using Y offset: 0)
;(Using Z offset: 0)
$32=1
G28 G91 Z0
G90

;(2D Contour5)
S12000 M4
G54
G0 X350.6 Y152.4
Z15
Z0
G1 Z-0.2 F333.3
Y261.462 F5000
G3 X345.6 Y266.462 I-5
G1 X60.8
G3 X55.8 Y261.462 J-5
G1 Y43.338
G3 X60.8 Y38.338 I5
G1 X345.6
G3 X350.6 Y43.338 J5
G1 Y152.4
G0 Z0
G28 G91 Z0
G90
G0 G91 X0 Y0
G90
M5
;G49
;(Disable the laser mode!)
$32=0
```

**Comment by quezadaminter** [comment]
@AustinSaintAubin, thank you for testing. I agree with you, I'll open a new issue.

**Comment by chamnit** [comment]
Be careful changing EEPROM settings in a program. See EEPROM issues in the wiki. It has to do with how EEPROM on the AVR works. Insert a G4 Px command before any settings change to ensure Grbls planner is empty and can safely alter the setting.

**Comment by quezadaminter** [comment]
FYI, I was able to obtain a solution to #598 (the issue opened by the comment above). Thank you guys for your help.
